### PR TITLE
Protein function prediction with GO - Part 2

### PR DESCRIPTION
**Pull request by aditya0by0**
- ### PR for the Issue #36 

Note: The above issue will be implemented in 2 PRs:
 - #39 (Merged)
 - #57 

Tasks
- Minimal dataset implementation: Build a dataset class that extracts proteins and labels from UniProtKB / GO and processes them into a dataset that can be used to train Electra

### CODE REVIEW

**Comment by aditya0by0** [comment]
Changes to implement from:  https://github.com/ChEB-AI/python-chebai/pull/39#issuecomment-2370456531

> Thanks for implementing this.
> 
> * For the sequence length: I would except the maximum sequence length to refer to the number of amino acids. That way, the same proteins are included in the dataset for a given sequence length, no matter the encoding.
> * Separate `tokens.txt` files for each n-gram: Definitely, since they have different sets of tokens (tokens always have length n for each n-gram). This should happen automatically if you change the `name` property of the reader.
> * Vocabulary size: That is easy to fix: Simply don't use a pretrained model. Since the pretraining has been done on SMILES, it makes no sense to use that model for protein sequences. (Maybe we will do pretraining for protein sequences in the future, then we will have to pretrain a model with `vocab_size=8000`)
> 
> I will merge this so we can use the classes for other PRs. Please open a new PR for this branch if you have new changes.

**Comment by aditya0by0** [comment]
Config:
```bash
class_path: chebai.preprocessing.datasets.go_uniprot.GOUniProtOver250
init_args:
  go_branch: "BP"
  reader_kwargs: {n_gram: 3}
```

**Comment by aditya0by0** [comment]
I have completed the changes suggested in our last meeting. Please review.

File: chebai/preprocessing/datasets/go_uniprot.py [hunk 1]
```diff
@@ -333,7 +341,18 @@ def _graph_to_raw_dataset(self, g: nx.DiGraph) -> pd.DataFrame:
...
+            lambda go_ids: list(
+                itertools.chain.from_iterable(
+                    [
+                        [go_id] + list(g.predecessors(go_id))
+                        for go_id in go_ids
+                        if go_id in g.nodes
+                    ]
+                )
+            )
+        )
```
Comment (aditya0by0) [inline chebai/preprocessing/datasets/go_uniprot.py:355]: ### Duplicate GO IDs Handling

To ensure we handle duplicate GO IDs properly, we should utilize a `set` to avoid including any GO ID more than once in our list of IDs. This approach will also help us account for common ancestors only once.

Additionally, this adjustment is crucial for the `select_classes` method to work properly, where we need to ensure that a single protein can annotate a specific GO ID only once. 


The necessity of this change is highlighted by the failure of the test case `TestGOUniProtDataExtractor.test_graph_to_raw_dataset`, which indicates that our current implementation may allow duplicates.

### Proposed Solution:
We can modify the code as follows:

```python
data_df["go_ids"] = data_df["go_ids"].apply(
    lambda go_ids: list(
        set(
            itertools.chain.from_iterable(
                [
                    [go_id] + list(g.predecessors(go_id))
                    for go_id in go_ids
                    if go_id in g.nodes
                ]
            )
        )
    )
)
```

This will ensure that we have a unique list of GO IDs and their ancestors for each entry, preventing any potential issues with duplicate annotations.

Comment (sfluegel05) [reply chebai/preprocessing/datasets/go_uniprot.py]: You are right. If a superclass appear several times, the list would count it twice instead of only once. That was a mistake on my part. Thanks for spotting this.
I replaced the `list` with your solution.

**Comment by aditya0by0** [comment]
> The next steps here are:
> 
> * Pretraining: add filter for sequence length as hyperparameter
> * merge the feature branch into the dev branch

Merging this branch, as suggested in comment https://github.com/ChEB-AI/python-chebai/issues/36#issuecomment-2447698109

A new PR with same branch, will be created for the rest of the changes
